[FSDP][1/N] Add `_get_fsdp_states()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90862 [FSDP][3/N] Move `fsdp_modules(root_only=True)` -> `_get_fsdp_root_states()`
* #90861 [FSDP][2/N] Move `fsdp_modules(root_only=False)` -> `_get_fsdp_states()`
* #90864 [FSDP][Easy] Rename `entry` -> `fsdp_module` to be more descriptive
* **#90860 [FSDP][1/N] Add `_get_fsdp_states()`**

- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - ~~I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.~~ Due to circular import issues, I think I should still put it in `_common_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

Commit:
[FSDP][1/N] Add `_get_fsdp_states()`

[ghstack-poisoned]

Commit:
Update on "[FSDP][1/N] Add `_get_fsdp_states()`"


- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1249,3 +1260,26 @@ def _cast_buffers_to_dtype_and_device(
+    for submodule in module.modules():
+        optional_state = _get_module_fsdp_state(submodule)
+        if optional_state is not None and optional_state not in visited_fsdp_states:
+            visited_fsdp_states.add(optional_state)
```
do we need visited set if we just let fsdp_states be a set and convert back to a list?

awgu replied:
We need to preserve the ordering. Let me make that clear.

rohan-varma reviewed (approved):
Would it be useful to have a unittest?

Commit:
Update on "[FSDP][1/N] Add `_get_fsdp_states()`"


- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - ~~I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.~~ Due to circular import issues, I think I should still put it in `_common_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

[ghstack-poisoned]

Commit:
Update on "[FSDP][1/N] Add `_get_fsdp_states()`"


- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - ~~I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.~~ Due to circular import issues, I think I should still put it in `_common_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

[ghstack-poisoned]

awgu commented:
> Would it be useful to have a unittest?

I think for this particular PR in the stack, it is okay to not include a unit test, since we just nee to make sure `fsdp_modules` backward compatibility is preserved. For later PRs in the stack, we will naturally require that this function works in order for the composable path to work.

Commit:
Update on "[FSDP][1/N] Add `_get_fsdp_states()`"


- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - ~~I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.~~ Due to circular import issues, I think I should still put it in `_common_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

mrshenli commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -40,6 +40,29 @@ def _get_module_fsdp_state(module: nn.Module) -> Optional[_FSDPState]:
+        optional_state = _get_module_fsdp_state(submodule)
+        if optional_state is not None and optional_state not in visited_fsdp_states:
+            visited_fsdp_states.add(optional_state)
+            fsdp_states.append(optional_state)
```
do we care about the order here (curious as we are using set + list)?

awgu replied:
Yes. I want to preserve the depth-first order that `module.modules()` follows, which is why I have this function return a list (and only use the set for tracking visited-ness).
